### ISSUE TITLE
[RFC]: Use vLLM's `extract_hidden_states` system (and enable online training)

### ISSUE THREAD

**Issue by fynnsu** [vllm-project/speculators#335]
### Motivation.

vLLM recently merged https://github.com/vllm-project/vllm/pull/33736 which adds direct support in vLLM for generating model hidden states. This allows us to deprecate the `VllmHiddenStatesGenerator` and replace this system with direct calls to vLLM. This should simplify the user/test experience for speculators, as users will no longer need to use separate environments for different parts of the speculators library.

In addition, moving the hidden states generation into a separate vLLM process allows us to make use of api endpoints, which should enable on-demand hidden states generation, as required by online training. 

This RFC explores the changes required in speculators to support the `extract_hidden_states` system, and online training.

## Extract hidden states (brief overview)
The vllm pr above added three components. A dummy eagle3 style model that receives hidden states and inserts them into its KV cache, an exmaple KV Connector specifically designed for extracting hidden states, and a dummy spec decode method for facilitating the pathway. 

If the model is setup correctly (i.e. configured with `extract_hidden_states` spec decode method and `ExampleHiddenStatesConnector` kv connector), then vllm requests (to python api or openai endpoint) will also receive a filepath in the response, pointing to a file containing the hidden states and token ids for the request prompt.

There are also plans to add more performant hidden states connectors that don't write to disk, but when they are added the output is likely to contain the same two tensors.

### Dataset-level requirements
1. Token frequency statistics: 
    **Currently:** produced during preprocessing and cached to disk. 
    **Used to:** produce vocab mappings (i.e. which tokens are most frequently used and therefore should be kept). 
    **Note:** this can be collected using the full dataset or just a (reasonably large) sample, although full dataset is better.
2. Per sample tokenized sequence lengths: 
    **Currently:** produced during preprocessing and cached to disk
    **Used to:** determine efficient batch packings
    **Note:** We have a fallback that approximates the sequence lengths based on the disk size of hidden states files if this information doesn't exist. Approximate seq length is sufficient for training

### Sample-level requirements
1. Hidden states (for eagle3 input layers and last verifier layer)
    **Currently:** produced during data gen and stored in `data_{i}.pt` file
2. Token ids
    **Currently:** produced during data gen and stored in `data_{i}.pt` file
3. Assistant mask
    **Currently:** produced during data gen and stored in `data_{i}.pt` file. In order to produce these, we simulate the `.apply_chat_template` process with known dummy inputs and use that to develop an assistant mask template. Then we use the mask template to compute masks on real samples

## (Proposed) Offline training flow
Use the existing logic for offline generation, except instead of using `VllmHiddenStatesGenerator` make calls to the (preconfigured) vLLM server `completions/` endpoint (passing in `token_ids`). 

Change existing file format from (current):
```
# data_{i}.pt
 {
  "input_ids": [seq_len],
  "loss_mask": [seq_len],
  "hidden_states": [
    [seq_len, hidden_size],
    [seq_len, hidden_size],
    [seq_len, hidden_size],
     ...
  ],
}
```
to
```
# data_{i}.pt
 {
  "input_ids": [seq_len],
  "loss_mask": [seq_len],
  "hidden_states_path": str
}
```
where "hidden_states_path" stores the relative path from the hidden states save location (configured when setting up vLLM ExampleHiddenStatesConnector) to the file (should be just the filename, which is the request id vLLM used internally). 

Dataset-level requirements (token frequencies and seq lengths) are collected and saved to disk as normal. 

During training, the training process loads `data_{i}.pt`, and then the `hidden_states_path` file to get the sample-level requirements. 

## Online training flow
Runs a preprocessing step that is essentially the logic above, but skips the vLLM call. The data saved to disk is
```
# data_{i}.pt
 {
  "input_ids": [seq_len],
  "loss_mask": [seq_len],
}
```
i.e. same as above minus the "hidden_states_path". 

During training, the training process loads `data_{i}.pt`, detects the `hidden_states_path` is missing, and calls vLLM. vLLM returns the `hidden_states_path`, which gets loaded and then deleted.

## Hybrid training flow
The above system also enables a hybrid training approach, where:

- Some hidden states could already be cached to disk, load the existing file if it exists
- Instead of deleting the `hidden_states_path` file after loading it, we can instead update `data_{i}.pt` to include the path. This would allow online gen for first epoch, and cached offline mode for subsequent epochs.

## Future (more performant) hidden states connectors
There is already ongoing work / plans for more performant connectors that don't simply write to disk. e..g a Nixl Connector that transfers hidden states directly from vllm (gpus) to the training process (gpus).

These should still work well the processed online system. We'll still need to compute the assistant masks, as well as pre-compute the dataset-level requirements (token frequencies and seq lengths). The missing hidden states will simply be transferred to the dataloader in a more efficient manner when it requests them.

## Ram disks
We can already (potentially) improve the performance of this system by pointing the hidden states save path to a ramdisk/tmpdisk. A ram disk is just a way to setup a directory on linux so that the storage actually uses ram instead of a hard disk/ssd. The `/tmp` directory on linux systems is typically configured this way. 

The nice thing about this, is that from the perspective of programs it's just another filepath and doesn't require any special handling, while having more performant read/writes. For online generation, users could use a hard drive/ssd filepath for their `data_{i}.pt` while setting the hidden states path to a ramdisk (if their system is set up with one). Then the online vLLM generation will write/read the hidden states to/from ram.

**Comment by benchislett** [comment]
Understandably out-of-scope, but good features to have
